File: kotlin/src/main/kotlin/structural/decorator/MainDecorator.kt
package structural.decorator

internal interface Component {
    fun operation(): String
}

internal class ConcreteComponent : Component {
    override fun operation(): String {
        return "concrete operation"
    }
}

internal abstract class Decorator : Component {
    protected lateinit var component: Component
}

internal class ConcreteDecorator1 : Decorator {
    constructor(component: Component) {
        super.component = component
    }

    override fun operation(): String {
        return "ConcreteDecorator1 " + component.operation()
    }
}

internal class ConcreteDecorator2 : Decorator {
    constructor(component: Component) {
        super.component = component
    }

    override fun operation(): String {
        return "ConcreteDecorator2 " + component.operation()
    }
}

fun main(args: Array<String>) {
    val component: Component = ConcreteComponent()
    val concreteDecorator1 = ConcreteDecorator1(component)
    val concreteDecorator2 = ConcreteDecorator2(component)
    println(concreteDecorator1.operation())
    println(concreteDecorator2.operation())
}

File: python/structural/decorator/MainDecorator.py
from abc import ABC, abstractmethod

class Component(ABC):
    @abstractmethod
    def operation(self) -> str: pass


class ConcreteComponent(Component):
    def operation(self) -> str:
        return "concrete operation"


class Decorator(Component):
    _component: Component


class ConcreteDecorator1(Decorator):
    def __init__(self, component: Component):
        self._component = component

    def operation(self) -> str:
        return self.__class__.__name__ + self._component.operation()



class ConcreteDecorator2(Decorator):
    def __init__(self, component: Component):
        self._component = component

    def operation(self) -> str:
        return self.__class__.__name__ + self._component.operation()


def main():
    component = ConcreteComponent()
    concreteDecorator1 = ConcreteDecorator1(component)
    concreteDecorator2 = ConcreteDecorator2(component)
    print(concreteDecorator1.operation())
    print(concreteDecorator2.operation())

main()